### ISSUE TITLE
proc-macro crate in workspace unexpectently exits

### ISSUE THREAD

**Issue by Javagedes** [xd009642/tarpaulin#1642]
**Describe the bug**
Workspaces with a proc-macro crate fail to execute tests (The failure happens when execution of the test exe begins, but before any test starts) with the following error:

```
2024-10-28T15:58:05.678629Z  INFO cargo_tarpaulin::process_handling: running D:\a\rust-ci-testing\rust-ci-testing\target\debug\deps\rust_proc_macro-6ef64dc17f0828ce.exe
2024-10-28T15:58:05.678931Z  INFO cargo_tarpaulin::process_handling: Setting LLVM_PROFILE_FILE
2024-10-28T15:58:05.686895Z ERROR cargo_tarpaulin: Test failed during run
Error: "Test failed during run"
Error: Process completed with exit code 1.
```

I can only reproduce on a github action `windows-latest` runner (but 100% of the time). It does not happen on ubuntu-latest image (even with the `--engine llvm` flag) and it does not happen on windows locally. it fails with or without any tests actually in the crate. It fails with `engine llvm` and `--no-dead-code`

I tried adding tests to see if that helped any, but it continued to fail. I also tried adding tests as described in #326 but it continues to fail.

I am currently on:
`cargo-tarpaulin 0.31.2`
`1.80.0`

**To Reproduce**
I have set up a repo with a CI job that hits the issue: [Javagedes/rust-ci-testing](https://github.com/Javagedes/rust-ci-testing) with an example run here: [action/Run Tests](https://github.com/Javagedes/rust-ci-testing/actions/runs/11557651668/job/32168166257)

**Expected behavior**
I would expect the test harness to not unexpectedly fail without running any tests.

**Comment by xd009642** [comment]
Hmm that's super odd, could you add `--dump-traces` to the tarpaulin flags. Hopefully it's not so long that GHA gives up on printing the logs. 

I don't really use windows so this might take me some time to replicate (I need to find a windows machine and setup a dev env or find a friend who can help to get a quicker turnaround)

**Comment by nightkr** [comment]
I'm seeing [something similar](https://github.com/kube-rs/kube/pull/1652#issuecomment-2508027464) on Linux, both in CI (GitHub Actions) and locally (Arch).

**Comment by xd009642** [comment]
oh weird/interesting that's very helpful thanks :+1:

**Comment by nightkr** [comment]
Output from --dump-traces: [tarpaulin_20241129173337.json](https://github.com/user-attachments/files/17961611/tarpaulin_20241129173337.json)

**Comment by nightkr** [comment]
Of course, feel free to split this into a separate issue if it seems unrelated.

**Comment by jonasbb** [comment]
Locally (Rust 1.83, Fedora Linux, amd64) I see these messages right before the test failure.
```
2024-12-02T21:57:21.820615Z  INFO cargo_tarpaulin::process_handling::linux: Launching test
2024-12-02T21:57:21.820661Z  INFO cargo_tarpaulin::process_handling: running /home/jbushart/projects/serde_with/target/debug/deps/serde_with_macros-ae28cd5b53cb79aa
/home/jbushart/projects/serde_with/target/debug/deps/serde_with_macros-ae28cd5b53cb79aa: error while loading shared libraries: libstd-02dc84f49135b6b8.so: cannot open shared object file: No such file or directory
2024-12-02T21:57:23.109267Z ERROR cargo_tarpaulin: Test failed during run
Error: "Test failed during run"
```

I do not see the link message on GitHub Actions ([Source](https://github.com/jonasbb/serde_with/actions/runs/12091800848/job/33720648041#step:14:997)).

```
test result: ok. 79 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.25s

Error: "Test failed during run"
Error: Process completed with exit code 1.
```

Since this is a linking issue, this change sounds relevant: "The sysroot no longer contains the std dynamic library in its top-level lib/ dir." https://github.com/rust-lang/rust/pull/131188

**Comment by xd009642** [comment]
Oh one thing that another person with the issue has confirmed is that `--exclude <PROC_MACRO_CRATE>` also works to solve the issue as a stop gap

**Comment by dkaluza** [comment]
For me setting dynamic libraries path environment variable manually worked as a workaround:

```LD_LIBRARY_PATH="$(rustc --print=target-libdir)" cargo tarpaulin -o html```

Note that name of the env variable differs depending on operating system: https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths

**Comment by xd009642** [comment]
Thanks for the tip @dkaluza I've just done a PR to get tarpaulin to automate this process for people here https://github.com/xd009642/tarpaulin/pull/1680 if anyone wants to try it out (I'll be trying it later) :eyes:

**Comment by waltronix** [comment]
I tried it and the fix worked fine for me.

**Comment by Sewer56** [comment]
In my case, I do still experience issues with `proc_macro` compilation (on nightly), in this case, it's not possible to build `clap_builder`.

```
error: could not compile `proc-macro-crate` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
2025-01-16T19:29:37.684526Z ERROR cargo_tarpaulin: Failed to compile tests!
error: symbol `_RNCNvCsiot3wJO3568_16proc_macro_crate17target_dep_tables0B3_` is already defined
   --> /home/sewer/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/proc-macro-crate-3.2.0/src/lib.rs:385:78
    |
385 |     cargo_toml.get("target").into_iter().filter_map(Item::as_table).flat_map(|t| {
    |                                                                              ^^^


Error: "Failed to compile tests!\nerror: symbol `_RNCNvCsiot3wJO3568_16proc_macro_crate17target_dep_tables0B3_` is already defined\n   --> /home/sewer/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/proc-macro-crate-3.2.0/src/lib.rs:385:78\n    |\n385 |     cargo_toml.get(\"target\").into_iter().filter_map(Item::as_table).flat_map(|t| {\n    |                                                                              ^^^\n\n"
```

It's a bit problematic, because the `criterion` benchmarking library relies on it:

```
clap_builder v4.5.26
└── clap v4.5.26
    └── criterion v0.5.1
            [dev-dependencies]
```

I'm not sure if the issue is related but today's release, which seems to encompass the above workaround doesn't do the trick for me.
Most likely, the build issue is unrelated.

**Comment by xd009642** [comment]
@Sewer56 which OS are you using?

**Comment by Sewer56** [comment]
> [@Sewer56](https://github.com/Sewer56) which OS are you using?

x86-64 Linux, specifically CachyOS.
This is a Nightly Rust only issue.

The above issue also affects Windows and macOS; including non-x86 platforms, so it doesn't need any specialized setup, thankfully. 

There is a small chance of a build succeeding but 9/10 times it will fail. That said, I've been unable to pass a single CI run for about a week.

https://github.com/Reloaded-Project/devops-rust-test-and-coverage/actions

A GitHub Composite Action I wrote illustrates this best. There I have a cron job to test the workflow every 24 hours. As visible here, 5 days ago all runs started failing on nightly Rust.

I've only got to look into it today, as I was incredibly busy (as there's always the chance the dependency itself is at fault), however I noticrd today that the dependency version did not change, and stable toolchain still runs fine.

**Comment by xd009642** [comment]
Interesting, I've noticed a similar failure in tarpaulin CI start 5 days ago - but only on the nightly mac builds. I think it's unrelated or something a bit different so if you want to open a new  issue I'll close this one

**Comment by Sewer56** [comment]
Yup, separate issue, figured I'd bring it up though.
I'll get that sorted next morning as I'm practically asleep right now. (Just had to wake up briefly)

**Comment by Sewer56** [comment]
Alright, woke up, and typed up a simple issue here:
- https://github.com/xd009642/tarpaulin/issues/1681

With some extra details/clarity too.
This one can probably (indeed) be closed.
